mdtk/downloaders.py

```diff
@@ -0,0 +1,233 @@
+"""
+Classes to download data from different source. Each class gets data from one
+source. The base data to get is midi. To contribute a new dataset, create a
+new class which extends DataDownloader, and write an accompanying test in
+./tests/test_downloads.py
+"""
+import os
+import sys
+import shutil
+import urllib
+import warnings
+import zipfile
+import glob
+from tqdm import tqdm
+
+
+
+USER_HOME = os.path.expanduser('~')
+DEFAULT_CACHE_PATH = os.path.join(USER_HOME, '.mdtk_cache')
+
+
+
+def download_file(source, dest, verbose=True, overwrite=None):
+        """Get a file from a url and save it locally"""
+        if verbose:
+            print(f'Downloading {source} to {dest}')
+        if os.path.exists(dest):
+            if overwrite is None:
+                warnings.warn(f'WARNING: {dest} already exists, not '
+                              'downloading', category=UserWarning)
+                return
+            if not overwrite:
+                raise OSError(f'{dest} already exists')
+        try:
+            urllib.request.urlretrieve(source, dest)
+        except urllib.error.HTTPError as e:
+            print(f'Url {source} does not exist', file=sys.stderr)
+            raise e
+
+
+def make_directory(path, overwrite=None, make_parents=True, verbose=True):
```

Comment:
I use os.makedirs(dir, exist=True) elsewhere in midi.py. Do we have a specific use case for this util function being needed?

Reply:
Yes, it's just the handling of whether the destination exists or not.